File: packages/java/ygo74-agent-runtime/src/main/java/com/ygo74/agentruntime/routing/DispatcherImpl.java
package com.ygo74.agentruntime.routing;

import com.ygo74.agentruntime.domains.contracts.ExchangeModels.StandardExchangeRequest;
import com.ygo74.agentruntime.domains.contracts.ExchangeModels.StandardExchangeResponse;

import java.util.function.Function;

public final class DispatcherImpl implements Dispatcher {
    @Override
    public StandardExchangeResponse dispatch(StandardExchangeRequest request, Function<String, UseCaseHandler> resolver) {
        return resolver.apply(request.routeKey()).handle(request);
    }
}

File: packages/python/ygo74/agent_runtime/routing/dispatcher_impl.py
from collections.abc import Callable

from ygo74.agent_runtime.domains.contracts.exchange_models import StandardExchangeRequest, StandardExchangeResponse


class DispatcherImpl:
    def dispatch(self, request: StandardExchangeRequest, resolver: Callable[[str], Callable[[StandardExchangeRequest], StandardExchangeResponse]]) -> StandardExchangeResponse:
        handler = resolver(request.route_key)
        return handler(request)
